src/common/KernelBase.hpp

```diff
@@ -206,81 +209,166 @@ class KernelBase
+
+  DataSpace getDataSpace(VariantID vid) const
+  {
+    switch ( vid ) {
+
+      case Base_Seq :
+      case Lambda_Seq :
+      case RAJA_Seq :
+        return getSeqDataSpace();
+
+      case Base_OpenMP :
+      case Lambda_OpenMP :
+      case RAJA_OpenMP :
+        return getOmpDataSpace();
+
+      case Base_OpenMPTarget :
+      case RAJA_OpenMPTarget :
+        return getOmpTargetDataSpace();
+
+      case Base_CUDA :
+      case Lambda_CUDA :
+      case RAJA_CUDA :
+        return getCudaDataSpace();
+
+      case Base_HIP :
+      case Lambda_HIP :
+      case RAJA_HIP :
+        return getHipDataSpace();
+
+      case Kokkos_Lambda :
+        return getKokkosDataSpace();
+
+      default:
+        throw std::invalid_argument("getDataSpace : Unknown variant id");
+    }
+  }
+
+  DataSpace getHostAccessibleDataSpace(VariantID vid) const
+  {
+    return hostAccessibleDataSpace(getDataSpace(vid));
+  }
+
+  template <typename T>
+  void allocData(DataSpace dataSpace, T& ptr, int len)
+  {
+    rajaperf::allocData(dataSpace,
+        ptr, len, getDataAlignment());
+  }
+  template <typename T>
+  void deallocData(DataSpace dataSpace, T& ptr)
+  {
+    rajaperf::deallocData(dataSpace, ptr);
+  }
+
+  template <typename T>
+  void allocData(T*& ptr, int len, VariantID vid)
+  {
+    rajaperf::allocData(getDataSpace(vid),
+        ptr, len, getDataAlignment());
+  }
+
+  template <typename T>
+  void allocAndInitData(T*& ptr, int len, VariantID vid)
+  {
+    rajaperf::allocAndInitData(getDataSpace(vid),
+        ptr, len, getDataAlignment());
+  }
+
+  template <typename T>
+  void allocAndInitDataConst(T*& ptr, int len, T val, VariantID vid)
+  {
+    rajaperf::allocAndInitDataConst(getDataSpace(vid),
+        ptr, len, getDataAlignment(), val);
+  }
+
   template <typename T>
-  void allocData(T*& ptr, int len,
-                 VariantID vid)
+  void allocAndInitDataRandSign(T*& ptr, int len, VariantID vid)
   {
-    rajaperf::detail::allocData(ptr, len, getDataAlignment(), vid);
+    rajaperf::allocAndInitDataRandSign(getDataSpace(vid),
+        ptr, len, getDataAlignment());
   }
 
   template <typename T>
-  void allocAndInitData(T*& ptr, int len,
-                        VariantID vid)
+  void allocAndInitDataRandValue(T*& ptr, int len, VariantID vid)
   {
-    rajaperf::detail::allocAndInitData(ptr, len, getDataAlignment(), vid);
+    rajaperf::allocAndInitDataRandValue(getDataSpace(vid),
+        ptr, len, getDataAlignment());
   }
 
   template <typename T>
-  void allocAndInitDataConst(T*& ptr, int len, T val,
-                             VariantID vid)
+  rajaperf::AutoDataMover<T> allocSetupData(T*& ptr, int len, VariantID vid)
   {
-    rajaperf::detail::allocAndInitDataConst(ptr, len, getDataAlignment(), val, vid);
+    rajaperf::allocData(getHostAccessibleDataSpace(vid),
+        ptr, len, getDataAlignment());
+    return {getDataSpace(vid), getHostAccessibleDataSpace(vid), ptr, len, getDataAlignment()};
   }
```

Comment:
But if there is any data initialization in the Kernel's setup method then you use the SetupData alloc methods which allocate memory that is accessible on the host, then you do setup, then the AutoDataMover moves it back at the end of the setup method. Maybe AutoDataMover is overly complicated, but I wanted the data to always be moved to the right DataSpace.